gcc-13/14 c++23 build errors

cmake (command from CI config):

```
cmake -DCMAKE_BUILD_TYPE=Release  -DBUILD_SHARED_LIBS=ON \
        -DCMAKE_CXX_STANDARD=23 -DFMT_DOC=OFF \
        -DCMAKE_CXX_VISIBILITY_PRESET=hidden -DCMAKE_VISIBILITY_INLINES_HIDDEN=ON \
        -DFMT_PEDANTIC=ON
```

error:

```
[  1%] Building CXX object test/gtest/CMakeFiles/gtest.dir/gmock-gtest-all.cc.o
[  2%] Building CXX object test/CMakeFiles/nolocale-test.dir/__/src/format.cc.o
[  3%] Building CXX object test/CMakeFiles/noexception-test.dir/__/src/format.cc.o
[  5%] Building CXX object CMakeFiles/fmt.dir/src/format.cc.o
[  6%] Linking CXX shared library libnolocale-test.so
[  6%] Built target nolocale-test
[  7%] Building CXX object test/CMakeFiles/noexception-test.dir/noexception-test.cc.o
In file included from /usr/include/stdio.h:43,
                 from /home/runner/work/fmt/fmt/include/fmt/base.h:17,
                 from /home/runner/work/fmt/fmt/include/fmt/format.h:41,
                 from /home/runner/work/fmt/fmt/include/fmt/format-inl.h:23,
                 from /home/runner/work/fmt/fmt/src/format.cc:8:
In function ‘int putc_unlocked(int, FILE*)’,
    inlined from ‘void fmt::v11::detail::glibc_file<F>::init_buffer() [with F = _IO_FILE]’ at /home/runner/work/fmt/fmt/include/fmt/format-inl.h:1515:18,
    inlined from ‘fmt::v11::detail::file_print_buffer<F, typename std::enable_if<fmt::v11::detail::has_flockfile<F>::value, void>::type>::file_print_buffer(F*) [with F = _IO_FILE]’ at /home/runner/work/fmt/fmt/include/fmt/format-inl.h:1665:22,
    inlined from ‘void fmt::v11::vprint(FILE*, string_view, format_args)’ at /home/runner/work/fmt/fmt/include/fmt/format-inl.h:1724:48:
/usr/include/x86_64-linux-gnu/bits/stdio.h:103:10: warning: null pointer dereference [-Wnull-dereference]
  103 |   return __putc_unlocked_body (__c, __stream);
      |          ^~~~~~~~~~~~~~~~~~~~
In function ‘int putc_unlocked(int, FILE*)’,
    inlined from ‘void fmt::v11::detail::glibc_file<F>::init_buffer() [with F = _IO_FILE]’ at /home/runner/work/fmt/fmt/include/fmt/format-inl.h:1515:18,
    inlined from ‘fmt::v11::detail::file_print_buffer<F, typename std::enable_if<fmt::v11::detail::has_flockfile<F>::value, void>::type>::file_print_buffer(F*) [with F = _IO_FILE]’ at /home/runner/work/fmt/fmt/include/fmt/format-inl.h:1665:22,
    inlined from ‘void fmt::v11::vprint(FILE*, string_view, format_args)’ at /home/runner/work/fmt/fmt/include/fmt/format-inl.h:1724:48,
    inlined from ‘void fmt::v11::vprint(string_view, format_args)’ at /home/runner/work/fmt/fmt/include/fmt/format-inl.h:1736:9:
/usr/include/x86_64-linux-gnu/bits/stdio.h:103:10: warning: null pointer dereference [-Wnull-dereference]
  103 |   return __putc_unlocked_body (__c, __stream);
      |          ^~~~~~~~~~~~~~~~~~~~
[  8%] Building CXX object CMakeFiles/fmt.dir/src/os.cc.o
In file included from /usr/include/stdio.h:43,
                 from /home/runner/work/fmt/fmt/include/fmt/base.h:17,
                 from /home/runner/work/fmt/fmt/include/fmt/format.h:41,
                 from /home/runner/work/fmt/fmt/include/fmt/format-inl.h:23,
                 from /home/runner/work/fmt/fmt/src/format.cc:8:
In function ‘int putc_unlocked(int, FILE*)’,
    inlined from ‘void fmt::v11::detail::glibc_file<F>::init_buffer() [with F = _IO_FILE]’ at /home/runner/work/fmt/fmt/include/fmt/format-inl.h:1515:18,
    inlined from ‘fmt::v11::detail::file_print_buffer<F, typename std::enable_if<fmt::v11::detail::has_flockfile<F>::value, void>::type>::file_print_buffer(F*) [with F = _IO_FILE]’ at /home/runner/work/fmt/fmt/include/fmt/format-inl.h:1665:22,
    inlined from ‘void fmt::v11::vprint(FILE*, string_view, format_args)’ at /home/runner/work/fmt/fmt/include/fmt/format-inl.h:1724:48:
/usr/include/x86_64-linux-gnu/bits/stdio.h:103:10: error: null pointer dereference [-Werror=null-dereference]
  103 |   return __putc_unlocked_body (__c, __stream);
      |          ^~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
make[2]: *** [CMakeFiles/fmt.dir/build.make:76: CMakeFiles/fmt.dir/src/format.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:163: CMakeFiles/fmt.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

```

--- a/.github/workflows/linux.yml
+++ b/.github/workflows/linux.yml
@@ -13,6 +13,7 @@ jobs:
         cxx: [g++-4.9, g++-10, clang++-9]
         build_type: [Debug, Release]
         std: [11]
+        shared: [""]
         include:
           - cxx: g++-4.9
             install: sudo apt install g++-4.9
@@ -54,7 +55,11 @@ jobs:
             std: 20
             cxxflags: -stdlib=libc++
             install: sudo apt install libc++-11-dev libc++abi-11-dev
-          - shared: -DBUILD_SHARED_LIBS=ON
+          - cxx: g++-11
+            build_type: Release
+            std: 23
+            install: sudo apt install g++-11
+            shared: -DBUILD_SHARED_LIBS=ON
 
     steps:
     - uses: actions/checkout@a5ac7e51b41094c92402da3b24376905380afc29 # v4.1.6

--- a/.github/workflows/macos.yml
+++ b/.github/workflows/macos.yml
@@ -12,11 +12,15 @@ jobs:
         os: [macos-13, macos-14]
         build_type: [Debug, Release]
         std: [11, 17, 20]
+        shared: [""]
         exclude:
           - { os: macos-13, std: 11 }
           - { os: macos-13, std: 17 }
         include:
-          - shared: -DBUILD_SHARED_LIBS=ON
+          - os: macos-14
+            std: 23
+            build_type: Release
+            shared: -DBUILD_SHARED_LIBS=ON
 
     runs-on: '${{ matrix.os }}'
 

